### ISSUE TITLE
Gradient utilities incomplete

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2393]
## Description

Gradient utility functions in the training base module are incomplete.

## Location

- File: `shared/training/base.mojo`
- Lines: 244, 263

### 1. NaN/Inf Detection (Line 244)
```
# TODO: Implement NaN/inf detection (#34)
```

Need to detect numerical instability in gradients during training.

### 2. Gradient Norm Computation (Line 263)
```
# TODO: Implement gradient norm computation (#34)
```

Need to compute gradient norms for:
- Gradient clipping
- Training diagnostics
- Exploding gradient detection

## Acceptance Criteria

- [ ] Implement `has_nan_or_inf(tensor)` function
- [ ] Implement `compute_gradient_norm(parameters)` function
- [ ] Support L1 and L2 norms
- [ ] Efficient SIMD implementation where possible
- [ ] Tests for edge cases (empty tensors, large values)
- [ ] Integration with training loop

## Related

References Issue #34 in TODO comments.
